core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,711 @@
+                LongConsumer getSplitTimeRecorder,
+                Optional<int[]> bucketToPartitionMap,
+                Optional<BucketNodeMap> bucketNodeMap)
+        {
+            checkArgument(bucketNodeMap.isPresent() || fragment.getPartitionedSources().isEmpty(), "bucketNodeMap is expected to be set");
+            Map<PlanNodeId, SplitSource> splitSources = splitSourceFactory.createSplitSources(session, fragment);
+            return new HashDistributionTaskSource(
+                    splitSources,
+                    getPartitionedExchangeSourceHandles(fragment, exchangeSourceHandles),
+                    getReplicatedExchangeSourceHandles(fragment, exchangeSourceHandles),
+                    splitBatchSize,
+                    getSplitTimeRecorder,
+                    bucketToPartitionMap,
+                    bucketNodeMap,
+                    fragment.getPartitioning().getConnectorId());
+        }
+
+        public HashDistributionTaskSource(
+                Map<PlanNodeId, SplitSource> splitSources,
+                Multimap<PlanNodeId, ExchangeSourceHandle> partitionedExchangeSourceHandles,
+                Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles,
+                int splitBatchSize,
+                LongConsumer getSplitTimeRecorder,
+                Optional<int[]> bucketToPartitionMap,
+                Optional<BucketNodeMap> bucketNodeMap,
+                Optional<CatalogName> catalogRequirement)
+        {
+            this.splitSources = ImmutableMap.copyOf(requireNonNull(splitSources, "splitSources is null"));
+            this.partitionedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(partitionedExchangeSourceHandles, "partitionedExchangeSourceHandles is null"));
+            this.replicatedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(replicatedExchangeSourceHandles, "replicatedExchangeSourceHandles is null"));
+            this.splitBatchSize = splitBatchSize;
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+            this.bucketToPartitionMap = requireNonNull(bucketToPartitionMap, "bucketToPartitionMap is null");
+            this.bucketNodeMap = requireNonNull(bucketNodeMap, "bucketNodeMap is null");
+            checkArgument(splitSources.isEmpty() || bucketNodeMap.isPresent(), "splitToBucketFunction must be set when split sources are present");
+            this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            if (finished || closed) {
+                return ImmutableList.of();
+            }
+
+            Map<Integer, Multimap<PlanNodeId, Split>> partitionToSplitsMap = new HashMap<>();
+            Map<Integer, HostAddress> partitionToNodeMap = new HashMap<>();
+            if (!splitSources.isEmpty()) {
```

Comment:
What is the case for HASH distributed stage which has splitsources (TS nodes). Is that possible?

Reply:
For example join of a bucketed table with a non bucketed table (only one table has to be shuffled).

Reply:
You can skip `if (!splitSources.isEmpty()) {`.  Iterating empty map is not a problem.